proposed/http-handlers/request-handlers-meta.md

```diff
@@ -368,6 +370,240 @@ executing that handler. If unable to route to a handler, it would instead
+    {
+        return $this->middleware->process($request, $this->nextHandler);
+    }
+}
+
+// Create a response prototype to return if no middleware can produce a response
+// on its own. This could be a 404, 500, or default page.
+$responsePrototype = (new Response())->withStatus(404);
+$innerHandler = new class ($responsePrototype) implements RequestHandlerInterface {
+    private $responsePrototype;
+
+    public function __construct(ResponseInterface $responsePrototype)
+    {
+        $this->responsePrototype = $responsePrototype;
+    }
+
+    public function handle(ServerRequestInterface $request) : ResponseInterface
+    {
+        return $this->responsePrototype;
+    }
+};
+
+$layer1 = new DecoratedRequestHandler(new RoutingMiddleware(), $innerHandler);
+$layer2 = new DecoratedRequestHandler(new AuthorizationMiddleware(), $layer1);
+
+$response = $layer2->handle(ServerRequestFactory::fromGlobals());
+```
+
+Similar to the queue-based middleware, request handlers serve two purposes in
+this system:
+
+- Producing a fallback response if no other layer does.
+- Dispatching middleware.
+
+#### Reusable Middleware Examples
+
+In the examples above, we have two middleware composed in each. In order for
+these to work in either situation, we need to write them such that they interact
+appropriately.
+
+Generally speaking, implementors of middleware should follow these guidelines:
+
+- Test the request for a required condition. If it does not satisfy that
+  condition, use a composed prototype response or a composed response factory
+  to generate and return a response.
+
+- If pre-conditions are met, delegate creation of the response to the provided
+  request handler, optionally providing a "new" request by manipulating the
+  provided request (e.g., `$handler->handle($request->withAttribute('foo',
+  'bar')`).
+
+- Either pass the response returned by the request handler unaltered, or provide
+  a new response by manipulating the one returned (e.g., `return
+  $response->withHeader('X-Foo-Bar', 'baz')`).
+
+The `AuthorizationMiddleware` is one that will exercise all three of these guidelines:
+
+- If authorization is required, but the request is not authorized, it will use a
+  composed prototype response to produce an "unauthorized" response.
+- If authorization is not required, it will delegate the request to the handler
+  without changes.
+- If authorization is required and the request is authorized, it will delegate
+  the request to the handler, and sign the response returned based on the request.
+
+```php
+class AuthorizationMiddleware implements MiddlewareInterface
+{
+    private $authorizationMap;
+    private $responsePrototype;
+
+    public function __construct(
+        AuthorizationMap $authorizationMap,
+        ResponseInterface $responsePrototype
+    ) {
+        $this->authorizationMap = $authorizationMap;
+        $this->responsePrototype = $responsePrototype;
```

Comment:
I'm not sure why we need a response prototype here. I think it would make more sense of the `AuthorizationMap` was responsible for generating/manipulating the response. That would simplify the example and keep the focus on the middleware.

Reply:
Agreed; making that change now.